src/cerebellum.py

```diff
@@ -0,0 +1,114 @@
+"""Pathfinding Module"""
+import pdb
+from queue import Queue
+from typing import List, Optional
+from src.snake import Snake
+from src.board import Board
+from src.coordinate import Coordinate
+
+
+class Cerebellum:
+    """Find paths to given coordinates."""
+    def __init__(self, brain, selected_algorithm: Optional[str]) -> None:
+        self.selected_algorithm: str = selected_algorithm if selected_algorithm is not None else 'breadth_first'
+        self.brain = brain
+        self.me = self.brain.me
+        self.board: Board = self.brain.board
+
+    def get_path(self, coordinate: Optional[Coordinate] = None, coordinates: Optional[List[Coordinate]] = None) -> List[Coordinate]:
+        """Find path to given coordinate/coordinates based on pathfinder's selected algorithm."""
+
+        if coordinates is None:
+            if coordinate is None:
+                return []
+            coordinates = [coordinate]
+
+        path: List[Coordinate] = []
+
+        if self.selected_algorithm == 'breadth_first':
+            path = self.__get_breadth_first_path(coordinates)
+
+#        elif self.selected_algorithm == 'djikstra\'s':
+#            path = self.__get_djikstras_path(coordinates)
+
+#        elif self.selected_algorithm == 'a*':
+#            path = self.__get_a_star_path(coordinates)
+
+        return path
+
+    def __get_breadth_first_path(self, coordinates: List[Coordinate]) -> List[Coordinate]:
+        """Find path using breadth-first algorithm."""
+
+        #pick location from grid (self.me.head)
+        #expand that location by looking at neighbours
+        #any unvisited neighbours are added to "unexplored" and "visited" set
+
+        for coord in coordinates:
+            print(coord)
+
+        start = self.me.head
+
+        unexplored: Queue = Queue()
+        unexplored.put(start)
+        came_from: dict = {}
+        came_from[str(start)] = None
+
+        collision_coordinates = [coordinate for snake in self.board.snakes for coordinate in snake.coordinates]
+
+        #remove this snake's tail from array of "invalid moves"
+        tail = self.brain.me.coordinates[-1]
+        if tail in collision_coordinates:
+            collision_coordinates.remove(tail)
+
+
+        matched = False
+        while not unexplored.empty():
+            current = unexplored.get()
+
+            if any(coord.x == current.x and coord.y == current.y for coord in coordinates):
+                matched = True
+                break
+
+            valid_move_coordinatees = [coord for coord in current.get_neighbours() if self.board.is_coordinate_in_bounds(coord) and coord not in collision_coordinates]
```

Comment:
This is the one thing we will have trouble doing if we don't pass the board class in. Thoughts? We could do the math ourselves internally here... or board could pass in a list of collision coordinates that include edges?

Reply:
board doesn't have a reference to brain, so I'm just passing "me" and "board" into cerebellum instead of brain and as far as I can tell there's no circular dependency